### DISCUSSION TITLE
Thank you.

### DISCUSSION THREAD

**Discussion by duckduckdog** [dechamps/RudeWindowFixer#8] [General]
This was making me crazy. 

Have you heard of windhawk.net? I have a hunch you have plenty to add that doesn't justify a whole executable.

Thank you for not just making a fix, but packaging it so that it was painless to use.

**Comment by dechamps** [comment]
Thanks for pointing me to Windhawk, I didn't know about it. Impressive stuff!

That said… Windhawk is a Windows hook injection framework. While hooks are an extremely powerful approach and can be used to do pretty much anything, they have the downside of being brittle because they heavily rely on the undocumented internals of the code being hooked. [Monkey patching](https://en.wikipedia.org/wiki/Monkey_patch) in this way is extremely unsupported by the software vendor (in this case Microsoft), and can easily be broken by updates or incorrect assumptions.

Now, it is true that RudeWindowFixer also relies on undocumented Windows internals. But the crucial difference is, if RudeWindowFixer has a bug, the blast radius is very small as it is isolated in a single independent process - at worst, it simply doesn't work. In contrast, the blast radius of code monkeypatched into an application process is much worse - a bug can crash or corrupt the entire host process. When that process is `explorer.exe` (or the user's apps), things go downhill very quickly.

I'm not saying that hooking should never be used - there are some scenarios in which it is the only practical solution and nothing else will work. But that should only be a last resort solution due to the intrusive and brittle nature of the approach. Almost anything else is preferable.

This is why I don't believe it is necessarily a good idea to rewrite RudeWindowFixer around a hooking approach. The separate process approach, while less flexible and perhaps slightly less efficient, is nonetheless less intrusive, more robust and easier to implement and debug.

**Comment by Ghis1964** [comment]
Saw many post everywhere but no fix. Then landed on superuser-dot-com and found RudeWindowFixer (read everything about, and the Read Me file. Download and run) I just made a short-cut in the startup folder for it to run when logging-in etc. 

For my case, things got such that I had to uncheck/recheck the always-onTop pref in gadgets/widgets every 5 minutes. Even restarting the app wouldn't cure this bug. Yeah, it is a windows bug like you said. I'd say ever since Vista, in my experience. Would sometime happen on win7. Maybe twice a year. But in win10... meh.

It's been only 3 hours, but haven't seen any window not respecting the "AlwaysOnTop" of other window. (7 Waterfox windows 1000s of tabs, 1 Chrome window 12 tabs, GIMP, 6 win-explorer windows, notepad++ 40+ tabs, 3 Cmd windows, Vbox resting).

I could not comment on superuser (the 50 points krap yeah) so decided to comment here, and now I realize it was your post on superuser that I wanted to write this comment (lol)

Thank you
As the guys from "JunkyardDigg" would say: "I don't know if it's fixed, but it's different" lol

 EDIT:

Next day now. Not working. Back to the start... again :/
